### PR TITLE
Improve TP/SP LoRA performance and sync uncached text inputs

### PR DESCRIPTION
**Pull request by gazingstars123**
## Summary
- add TP/SP fused QKV and packed LoRA execution optimizations
- trim TP/SP diagnostics, share LoRA communication, and add a no-input-grad fast path for frozen context
- fix uncached TP text input synchronization across ranks and expose TP async overlap as an experimental UI option

## Notes
- includes the follow-up optimizer guard for offload/block-swap device mismatches
- leaves unrelated untracked local files out of the PR

## Summary by CodeRabbit

* **New Features**
  * UI TP/SP backend selector and enforced Sequence Parallel; new parallel runtime package enabling TP+SP training, packed-projection support, and FSDP helpers
  * Parallel checkpoint manifest format with coordinated rank-sharded save/resume and controlled multi-rank uploads

* **Bug Fixes**
  * Prevent duplicate epoch logs from DataLoader workers
  * Reconcile optimizer parameter devices with gradients before stepping

* **Improvements**
  * Centralized Flash Attention backend detection and clearer logging
  * Expanded TP/SP diagnostics, async overlap option, safer LoRA save/restore, server heartbeat to drop stale connections

* **Documentation**
  * Added optimization log recording active training optimizations and compatibility notes

### CODE REVIEW

**Comment by gazingstars123** [comment]
@coderabbitai full review

**Comment by gazingstars123** [comment]
@coderabbitai generate sequence diagram

**Comment by gazingstars123** [comment]
> ## Sequence Diagram for PR
> ```mermaid
> sequenceDiagram
>     participant UI as Web UI Server
>     participant Rank0 as Rank 0 Trainer
>     participant RankK as Rank K Trainer
>     participant TPModule as TP/SP Modules
>     participant LoRA as LoRA Handler
>     participant FS as FileSystem
>     participant Optimizer as Optimizer
> 
>     UI->>Rank0: Configure tp_backend, tp_degree, no_fuse_qkv, tp_async_overlap
>     Rank0->>TPModule: Initialize TP+SP with group wrappers
>     Rank0->>TPModule: Fuse QKV→PackedColumnParallel (if enabled)
>     
>     Note over Rank0,RankK: Training Forward Pass
>     Rank0->>TPModule: Broadcast uncached text inputs across ranks
>     RankK->>TPModule: Receive aligned text inputs
>     Rank0->>TPModule: Apply QKV fusion & padding
>     Rank0->>LoRA: Apply packed LoRA modules (prepare phase)
>     Rank0->>TPModule: Select Flash Attention backend
>     Rank0->>TPModule: Compute attention with RoPE padding
>     
>     Note over Rank0,RankK: Backward & Gradient Sync
>     Rank0->>LoRA: Gradient computation (apply phase)
>     Rank0->>TPModule: Synchronize LoRA grads across TP-sharded vs SP-partial replicated params
>     RankK->>TPModule: Receive gradient updates
>     Rank0->>Optimizer: Ensure optimizer param devices match grad devices
>     Rank0->>Optimizer: Step with aligned parameters
>     
>     Note over Rank0,RankK: Parallel Save/Resume
>     Rank0->>FS: Create state_root/ and rank_00000/ checkpoint
>     RankK->>FS: Write rank_0000K/ checkpoint in parallel
>     Rank0->>FS: Write tp_state.json manifest with parallel_size validation
>     Rank0->>FS: Write save-id marker & finalize root metadata
>     
>     Note over Rank0,RankK: Resume Flow
>     Rank0->>FS: Detect tp_state.json & validate parallel_size
>     RankK->>FS: Load only rank_0000K/ state
>     Rank0->>FS: Reconstruct root-level metadata (trim/pad gather)
>     Rank0->>UI: Resume complete, training continues
> ```

sure bud

**Comment by gazingstars123** [comment]
@coderabbitai full review

**Comment by gazingstars123** [comment]
@coderabbitai full review
